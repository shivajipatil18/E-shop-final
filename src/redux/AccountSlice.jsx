import { createSlice, createAsyncThunk } from '@reduxjs/toolkit';
import axios from 'axios';


export const fetchUser = createAsyncThunk('account/fetchUser', async (id, { rejectWithValue }) => {
  try {
    const res = await axios.get(`http://localhost:3001/users/${id}`);
    return res.data;
  } catch (err) {
    return rejectWithValue(err.response?.data || 'Failed to fetch user');
  }
});

export const updateUserInfo = createAsyncThunk(
  'account/updateUserInfo',
  async ({ id, data }, { rejectWithValue }) => {
    try {
      const res = await axios.patch(`http://localhost:3001/users/${id}`, data);
      return res.data;
    } catch (err) {
      return rejectWithValue(err.response?.data || 'Failed to update info');
    }
  }
);

// Change password
export const changePassword = createAsyncThunk(
  'account/changePassword',
  async ({ id, currentPassword, newPassword }, { rejectWithValue, getState }) => {
    try {
      const { account } = getState();
      if (account.user.password !== currentPassword) {
        return rejectWithValue('Current password is incorrect.');
      }
      const res = await axios.patch(`http://localhost:3001/users/${id}`, { password: newPassword });
      return res.data;
    } catch (err) {
      return rejectWithValue(err.response?.data || 'Password update failed');
    }
  }
);

const accountSlice = createSlice({
  name: 'account',
  initialState: {
    user: null,
    status: 'idle',
    error: null,
  },
  reducers: {
    clearAccountState: (state) => {
      state.user = null;
      state.status = 'idle';
      state.error = null;
    },
  },
  extraReducers: (builder) => {
    builder
      .addCase(fetchUser.pending, (state) => {
        state.status = 'loading';
      })
      .addCase(fetchUser.fulfilled, (state, action) => {
        state.status = 'succeeded';
        state.user = action.payload;
      })
      .addCase(fetchUser.rejected, (state, action) => {
        state.status = 'failed';
        state.error = action.payload;
      })
      .addCase(updateUserInfo.fulfilled, (state, action) => {
        state.user = { ...state.user, ...action.payload };
      }).addCase(updateUserInfo.rejected, (state, action) => {
        state.error = action.payload;
      })
      .addCase(changePassword.fulfilled, (state, action) => {
        state.user = { ...state.user, password: action.payload.password };
      }).addCase(changePassword.rejected, (state, action) => {
        state.status = 'failed';
        state.error = action.payload || action.error.message;
      });;
  },
});

export const { clearAccountState } = accountSlice.actions;
export default accountSlice.reducer;
