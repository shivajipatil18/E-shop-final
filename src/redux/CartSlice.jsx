import { createSlice } from "@reduxjs/toolkit";

let initialState = {
  products: [],
  totalQuantity: 0,
  totalPrice: 0,
  searchResults: [],
};
let storedCart ="";
if (typeof window !== "undefined") {
storedCart = window.localStorage.getItem("cart");
}
if (storedCart) {
  initialState = JSON.parse(storedCart);
}
const cartSlice = createSlice({
  name: "product",
  initialState,
  reducers: {
    searchProduct: (state, action) => {
      const query = action.payload.toLowerCase();
      state.searchResults = state.products.filter((product) =>
        product.title.toLowerCase().includes(query)
      );
    },
  
    addToCart(state, action) {
      const newItem = action.payload;
      const itemIndex = state.products.findIndex(
        (item) => item.id === newItem.id
      );
    
      if (itemIndex !== -1) {
        state.products[itemIndex].quantity++;
        state.products[itemIndex].totalPrice += newItem.price;
      } else {
        state.products.push({
          id: newItem.id,
          title: newItem.title,
          price: newItem.price,
          quantity: 1,
          totalPrice: newItem.price,
          image: newItem.image,
        });
      }
    
      state.totalQuantity = state.products.reduce((sum, item) => sum + item.quantity, 0);
      state.totalPrice = state.products.reduce((sum, item) => sum + item.quantity * item.price, 0)
    },
    
    removeFromCart(state, action) {
      const id = action.payload;
      const findItem = state.products.findIndex((item) => item.id === id);
      if (findItem !== -1) {
        state.totalPrice -= state.products[findItem].totalPrice;
        state.totalQuantity -= state.products[findItem].quantity;
        state.products = state.products.filter((item) => item.id !== id);
      }
    },
    increaseQuantity(state, action) {
      const id = action.payload;
      const findItem = state.products.findIndex((item) => item.id === id);
      if (findItem !== -1) {
        state.products[findItem].quantity++;
        state.products[findItem].totalPrice += state.products[findItem].price;
        state.totalQuantity++;
        state.totalPrice += state.products[findItem].price;
      }
    },
    decreaseQuantity(state, action) {
      const id = action.payload;
      const findItem = state.products.findIndex((item) => item.id === id);
      if (findItem !== -1 && state.products[findItem].quantity > 1) {
        state.products[findItem].quantity--;
        state.products[findItem].totalPrice -= state.products[findItem].price;
        state.totalQuantity--;
        state.totalPrice -= state.products[findItem].price;
      }
      
    },
    setCart: (state, action) => {
      return action.payload;
    },
    clearCart(state) {
      state.products = [];
    },
  },  
  
});
export const { addToCart, removeFromCart, increaseQuantity, decreaseQuantity,setCart,searchProduct,clearCart } =
  cartSlice.actions;
export default cartSlice.reducer;
