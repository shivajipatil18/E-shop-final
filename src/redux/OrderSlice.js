
import { createSlice, createAsyncThunk } from "@reduxjs/toolkit";
import axios, { all } from "axios";

 export const fetchAllOrders = createAsyncThunk('orders/fetchOrders', async () => {
   const response = await axios.get('http://localhost:5000/orders');   
    return response.data;

 });

export const saveOrder = createAsyncThunk("order/saveOrder", async (order, thunkAPI) => {
  try {
    const response = await axios.post("http://localhost:5000/orders", order);
      return response.data;
    
  } catch (error) {
    return thunkAPI.rejectWithValue(error.message);
  }
});
  export const fetchOrders = createAsyncThunk("order/fetchOrders", async (username, thunkAPI) => {
    try {
      const response = await axios.get(`http://localhost:5000/orders?username=${username}`); 
      return response.data;
    } catch (error) {
      return thunkAPI.rejectWithValue(error.message);
    }
  });
  export const updateOrderStatus = createAsyncThunk(
    "order/updateOrderStatus",
    async ({ orderId, newStatus }) => {
      const response = await fetch(`http://localhost:5000/orders/${orderId}/status`, { 
        method: "PUT",
        headers: {
          "Content-Type": "application/json",
        },
        body: JSON.stringify({ status: newStatus }),
      });
  console.log(response,orderId,newStatus);
      if (!response.ok) {
        throw new Error("Failed to update order status");
      }
  
      return await response.json(); 
    }
  );
  
const orderSlice = createSlice({
  name: "order",
  initialState: {
    orderDetails: [],
    loading: false,
    error: null,
    allOrders: [],
  },
  reducers: {
    setOrderDetails: (state, action) => {
      state.orderDetails = [action.payload];
    },
    clearOrderDetails: (state) => {
      state.orderDetails = [];
    },
  },
  extraReducers: (builder) => {
    builder
      .addCase(saveOrder.pending, (state) => {
        state.loading = true;
        state.error = null;
      })
      .addCase(saveOrder.fulfilled, (state, action) => {
        state.loading = false;
        state.orderDetails = [action.payload];
      })
      .addCase(saveOrder.rejected, (state, action) => {
        state.loading = false;
        state.error = action.payload;
      }) .addCase(fetchOrders.pending, (state) => {
        state.loading = true;
        state.error = null;
      })
      .addCase(fetchOrders.fulfilled, (state, action) => {
        state.loading = false;
        state.orderDetails = action.payload;
      })
      .addCase(fetchOrders.rejected, (state, action) => {
        state.loading = false;
        state.error = action.payload;
      })  .addCase(fetchAllOrders.pending, (state) => {
        state.loading = true;
        state.error = null;
      })
      .addCase(fetchAllOrders.fulfilled, (state, action) => {
        state.loading = false;
        state.allOrders = action.payload;
      })
      // .addCase(fetchAllOrders.rejected, (state, action) => {
      //   state.loading = false;
      //   state.error = action.payload || "Something went wrong";
      // })  .addCase(updateOrderStatus.pending, (state) => {
      //   state.loading = true;
      //   state.error = null;
      // })
      // .addCase(updateOrderStatus.fulfilled, (state, action) => {
      //   state.loading = false;
      //   const updatedOrder = action.payload;
      //   const index = state.allOrders.findIndex((order) => order.id === updatedOrder.orderId);
      //   if (index !== -1) {
      //     state.allOrders[index].status = updatedOrder.newStatus;
      //   }
      // })
      // .addCase(updateOrderStatus.rejected, (state, action) => {
      //   state.loading = false;
      //   state.error = action.payload || action.error.message;
      // });
  },
});
      
  


export const { setOrderDetails, clearOrderDetails ,} = orderSlice.actions;
export default orderSlice.reducer;
