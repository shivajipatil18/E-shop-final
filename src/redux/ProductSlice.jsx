import { createSlice, createAsyncThunk } from "@reduxjs/toolkit";
import axios from "axios";

// Fetch all products
export const fetchProducts = createAsyncThunk('products/fetchProducts', async () => {
  const response = await axios.get('https://fakestoreapi.com/products');
  return response.data;
});

// Add a new product
export const addProduct = createAsyncThunk('products/addProduct', async (product) => {
  const response = await axios.post('https://fakestoreapi.com/products', product);
  return {
    ...response.data,
    rating: { rate: 0, count: 0 }, 
  };
});


export const updateProduct = createAsyncThunk('products/updateProduct', async (product) => {
  const response = await axios.put(`https://fakestoreapi.com/products/${product.id}`, product);
  return response.data;
});


export const deleteProduct = createAsyncThunk('products/deleteProduct', async (id) => {
  await axios.delete(`https://fakestoreapi.com/products/${id}`);
  return id;
});


export const selectFilteredProducts = (state) => {
  const { products = [], category, price } = state.product || {};
  return products.filter((product) => {
    const matchCategory = category ? product.category === category : true;
    const matchPrice = product.price <= price;
    // const matchRating = typeof product.rating === 'number' ? product.rating >= minRating : false;

    return matchCategory && matchPrice //&& matchRating;
    
  });
};

const initialState = {
  products: [],
  searchTerm: "",
  filteredData: [],
  loading: false,
  error: null,
  category: "",
  price: 1000,
  // minRating: 4, 

};

const productSlice = createSlice({
  name: "product",
  initialState,
  reducers: {
    setCategory: (state, action) => {
      state.category = action.payload;
    },
    setPrice: (state, action) => {
      state.price = action.payload;
    },
    setSearchTerm: (state, action) => {
      state.searchTerm = action.payload;
    },
    setFilteredData: (state, action) => {
      state.filteredData = action.payload;
    },
    setFilteredProducts(state, action) {
      state.filteredProducts = action.payload;
    },
  },
  extraReducers: (builder) => {
    builder
      .addCase(fetchProducts.pending, (state) => {
        state.loading = true;
        state.error = null;
      })
      .addCase(fetchProducts.fulfilled, (state, action) => {
        state.loading = false;
        state.products = [...state.products, ...action.payload];
        state.filteredData = action.payload;
      })
      .addCase(fetchProducts.rejected, (state, action) => {
        state.loading = false;
        state.error = action.error.message;
      })
  
  },
});

export const { setCategory, setPrice, setSearchTerm, setFilteredData,setFilteredProducts } = productSlice.actions;
export default productSlice.reducer;