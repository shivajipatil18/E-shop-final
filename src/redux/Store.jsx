import { configureStore } from "@reduxjs/toolkit";

import productReducer from "./ProductSlice";
import cartReducer from "./CartSlice";
import authReducer from './authSlice';
import orderReducer from "./OrderSlice";
import accountReducer from "./AccountSlice"
import userReducer from "./UserSlice";
import { use } from "react";

const store = configureStore({
  reducer: {
    product: productReducer,
    cart: cartReducer,
    auth: authReducer,
    order: orderReducer,
    account: accountReducer,
    user: userReducer,
  },
});
export default store;