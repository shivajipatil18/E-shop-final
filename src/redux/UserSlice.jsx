import { createSlice, createAsyncThunk } from "@reduxjs/toolkit";
import axios from "axios";

export const fetchAllUsers = createAsyncThunk("user/fetchAllUsers", async (_, thunkAPI) => {
  try {
    const response = await axios.get("http://localhost:5000/users");
    return response.data;
  } catch (error) {
    return thunkAPI.rejectWithValue(error.message);
  }
});

const userSlice = createSlice({
  name: "user",
  initialState: {
    users: [],
    loading: false,
    error: null,
  },
  reducers: {},
  extraReducers: (builder) => {
    builder
      .addCase(fetchAllUsers.pending, (state) => {
        state.loading = true;
        state.error = null;
      })
      .addCase(fetchAllUsers.fulfilled, (state, action) => {
        state.loading = false;
        state.users = action.payload;
      })
      .addCase(fetchAllUsers.rejected, (state, action) => {
        state.loading = false;
        state.error = action.payload;
      });
  },
});

export default userSlice.reducer;
