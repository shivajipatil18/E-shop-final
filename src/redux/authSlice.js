import { createSlice, createAsyncThunk } from '@reduxjs/toolkit';
import axios from 'axios';

export const register = createAsyncThunk('auth/register', async ({ name, email, password }, { rejectWithValue }) => {
  try {
    const response = await axios.post('http://localhost:3001/users', {
      name,
      email,
      password,
    });
    return response.data;
  } catch (error) {
    return rejectWithValue(error.response.data);
  }
});

export const login = createAsyncThunk('auth/login', async ({ email, password }, { rejectWithValue }) => {
  try {
    const response = await axios.get(`http://localhost:3001/users?email=${email}&password=${password}`);
     console.log(response);
     if(response.data.length && response.data[0].email !== email){
       return [];
     }
     return response.data;
  
  } catch (error) {
    return rejectWithValue(error.response.data);
  }
});

const authSlice = createSlice({
  name: 'auth',
  initialState: {
    isAuthenticated: false,
    status: 'idle',
    error: null,
    user: null,
  },
  reducers: {
    logout: (state) => {
      state.isAuthenticated = false;
      state.user = null;
    },
  },
  extraReducers: (builder) => {
    builder
      .addCase(register.pending, (state) => {
        state.status = 'loading';
        state.error = null;
      })
      .addCase(register.fulfilled, (state, action) => {
        state.status = 'succeeded';
        state.isAuthenticated = true;
        state.user = action.payload;
        state.error = null;
      })
      .addCase(register.rejected, (state, action) => {
        state.status = 'failed';
        state.error = action.payload || 'Registration failed';
      })
      .addCase(login.pending, (state) => {
        state.status = 'loading';
        state.error = null;
      })
      .addCase(login.fulfilled, (state, action) => {
        state.status = 'succeeded';
        state.isAuthenticated = true;
        state.user = action.payload;
        state.error = null;
      })
      .addCase(login.rejected, (state, action) => {
        state.status = 'failed';
        state.error = action.payload || 'Login failed';
      });
  },
});

export const { logout } = authSlice.actions;
export default authSlice.reducer;