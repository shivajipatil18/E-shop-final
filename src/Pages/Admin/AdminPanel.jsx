import React, { useEffect } from "react";
import { useDispatch, useSelector } from "react-redux";
import { fetchAllOrders } from "../../redux/OrderSlice";
import { fetchAllUsers } from "../../redux/UserSlice";
import {
  updateProduct,
  deleteProduct,
  fetchProducts,
} from "../../redux/ProductSlice";

const AdminPanel = () => {
  const dispatch = useDispatch();

  const {
    allOrders,
    loading: orderLoading,
    error: orderError,
  } = useSelector((state) => state.order || {});

  const {
    users,
    loading: userLoading,
    error: userError,
  } = useSelector((state) => state.user || {});

  const {
    products,
    loading: productLoading,
    error: productError,
  } = useSelector((state) => state.product || {});

  useEffect(() => {
    dispatch(fetchAllOrders());
    dispatch(fetchAllUsers());
    dispatch(fetchProducts());
  }, [dispatch]);

  const handleUpdateProduct = async (productId, updatedData) => {
    try {
      await dispatch(updateProduct({ productId, updatedData })).unwrap();
      console.log("Product updated successfully");
      dispatch(fetchProducts()); // Refresh the product list
    } catch (error) {
      console.error("Error updating product:", error);
    }
  };

  const handleDeleteProduct = async (productId) => {
    try {
      await dispatch(deleteProduct(productId)).unwrap();
      console.log("Product deleted successfully");
      dispatch(fetchProducts()); 
    } catch (error) {
      console.error("Error deleting product:", error);
    }
  };

  if (orderLoading || userLoading || productLoading) {
    return (
      <div className="text-center py-20 text-lg text-gray-600">
        Loading data...
      </div>
    );
  }

  if (orderError || userError || productError) {
    return (
      <div className="text-center py-20 text-red-600">
        Error: {orderError || userError || productError}
      </div>
    );
  }

  if (!allOrders || allOrders.length === 0) {
    return (
      <div className="container mx-auto py-16 text-center">
        <h2 className="text-3xl font-bold text-gray-800 mb-4">
          No Orders Found
        </h2>
      </div>
    );
  }

  return (
    <div className="container mx-auto py-12 px-4 md:px-10">
      <h2 className="text-4xl font-bold mb-8 text-center text-gray-800">
        Admin Panel - All Orders
      </h2>

      {/* Orders Table */}
      <div className="overflow-x-auto">
        <table className="min-w-full bg-white border border-gray-200 shadow-md rounded-xl overflow-hidden">
          <thead className="bg-gray-100 text-gray-700 text-sm">
            <tr>
              <th className="px-4 py-3 border">Order #</th>
              <th className="px-4 py-3 border">Username</th>
              <th className="px-4 py-3 border">Date</th>
              <th className="px-4 py-3 border">Total Price</th>
              <th className="px-4 py-3 border">Payment</th>
              <th className="px-4 py-3 border">Address</th>
              <th className="px-4 py-3 border">Products</th>
              {/* <th className="px-4 py-3 border">Status</th>
              <th className="px-4 py-3 border">Actions</th> */}
            </tr>
          </thead>
          <tbody className="text-sm text-gray-800">
            {allOrders.map((order, index) => (
              <tr key={index} className="border-t hover:bg-gray-50">
                <td className="px-4 py-3 border text-center text-red-600 font-semibold">
                  #{order.orderNumber}
                </td>
                <td className="px-4 py-3 border text-center">
                  {order.username}
                </td>
                <td className="px-4 py-3 border text-center">{order.date}</td>
                <td className="px-4 py-3 border text-center text-green-600 font-bold">
                  ₹ {order.totalPrice}
                </td>
                <td className="px-4 py-3 border text-center">
                  <span
                    className={`inline-block px-2 py-1 rounded text-xs font-medium ${
                      order.paymentMethod === "Cash on Delivery"
                        ? "bg-yellow-100 text-yellow-700"
                        : "bg-blue-100 text-blue-700"
                    }`}
                  >
                    {order.paymentMethod}
                  </span>
                </td>
                <td className="px-4 py-3 border text-sm">
                  {order.shippingInformation?.address},{" "}
                  {order.shippingInformation?.city},{" "}
                  {order.shippingInformation?.zip}
                </td>
                <td className="px-4 py-3 border">
                  <div className="space-y-3">
                    {order.products.map((product, idx) => (
                      <div
                        key={`${product.id}-${idx}`}
                        className="flex justify-between items-center gap-3 border-b pb-2"
                      >
                        <div className="flex items-center gap-3">
                          <img
                            src={product.image}
                            alt={product.name}
                            className="w-10 h-10 object-contain border rounded"
                          />
                          <div>
                            <p className="font-medium">{product.name}</p>
                            <p className="text-xs text-gray-500">
                              ₹{product.price} × {product.quantity}
                            </p>
                          </div>
                          {/* <td className="px-4 py-2 border text-center">
                            
                  <select
                    value={order.status || "Pending"}  // Use default value if status is missing
                    onChange={(e) => handleUpdateOrderStatus(order.id, e.target.value)}
                    className="p-1 border rounded"
                  >
                    <option value="Pending">Pending</option>
                    <option value="Shipped">Shipped</option>
                    <option value="Cancelled">Cancelled</option>
                  </select>
                </td> */}

                        </div>
                        <div>
                          {/* <button
                            onClick={() =>
                              handleUpdateProduct(product.id, {
                                name: product.name,
                                price: product.price,
                              })
                            }
                            className="text-blue-600 hover:text-blue-800 text-xs"
                          >
                            Update
                          </button>
                          <button
                            onClick={() => handleDeleteProduct(product.id)}
                            className="text-red-600 hover:text-red-800 ml-2 text-xs"
                          >
                            Delete
                          </button> */}
                        </div>
                      </div>
                    ))}
                  </div>
                </td>
              </tr>
            ))}
          </tbody>
        </table>
      </div>

      {/* Users Table */}
      <h2 className="text-3xl font-bold mt-16 mb-6 text-center text-gray-800">
        All Registered Users
      </h2>
      <div className="overflow-x-auto">
        <table className="min-w-full bg-white border border-gray-200 shadow-md rounded-xl overflow-hidden">
          <thead className="bg-gray-100 text-gray-700 text-sm">
            <tr>
              <th className="px-4 py-3 border">User ID</th>
              <th className="px-4 py-3 border">Username</th>
              <th className="px-4 py-3 border">Email</th>
              <th className="px-4 py-3 border">Role</th>
            </tr>
          </thead>
          <tbody className="text-sm text-gray-800">
            {users.map((user, index) => (
              <tr
                key={`${user.id}-${index}`}
                className="border-t hover:bg-gray-50"
              >
                <td className="px-4 py-3 border text-center">{user.id}</td>
                <td className="px-4 py-3 border text-center">
                  {user.username}
                </td>
                <td className="px-4 py-3 border text-center">{user.email}</td>
                <td className="px-4 py-3 border text-center capitalize">
                  {user.role || "customer"}
                </td>
              </tr>
            ))}
          </tbody>
        </table>
      </div>
    </div>
  );
};

export default AdminPanel;
