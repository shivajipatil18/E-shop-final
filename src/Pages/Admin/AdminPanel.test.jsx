import React from "react";
import { mount } from "enzyme";
// Removed unused import of configureMockStore
import thunk from "redux-thunk";
import { Provider } from "react-redux";
import AdminPanel from "../../Pages/Admin/AdminPanel";

import * as orderActions from "../../redux/OrderSlice";
import * as userActions from "../../redux/UserSlice";
import * as productActions from "../../redux/ProductSlice";

import { configureStore } from "@reduxjs/toolkit";

const createMockStore = (preloadedState) =>
  configureStore({
    reducer: () => preloadedState,
    middleware: (getDefaultMiddleware) => getDefaultMiddleware(),
  });
console.log("🛠️ createMockStore:", typeof createMockStore);
describe("AdminPanel Component", () => {
  let store;
  let wrapper;

  const mockState = {
    order: {
      allOrders: [
        {
          orderNumber: "1234",
          username: "john_doe",
          date: "2025-04-20",
          totalPrice: 2999,
          paymentMethod: "Cash on Delivery",
          shippingInformation: {
            address: "123 Main St",
            city: "Pune",
            zip: "411001",
          },
          products: [
            {
              id: 1,
              name: "Product 1",
              price: 1000,
              quantity: 2,
              image: "product1.jpg",
            },
          ],
        },
      ],
      loading: false,
      error: null,
    },
    user: {
      users: [
        { id: 1, username: "john_doe", email: "john@example.com", role: "admin" },
      ],
      loading: false,
      error: null,
    },
    product: {
      products: [],
      loading: false,
      error: null,
    },
  };
  store = createMockStore(mockState);

  beforeEach(() => {
    store = createMockStore(mockState);

    jest.spyOn(orderActions, "fetchAllOrders").mockImplementation(() => () => ({ type: "FETCH_ORDERS" }));
    jest.spyOn(userActions, "fetchAllUsers").mockImplementation(() => () => ({ type: "FETCH_USERS" }));
    jest.spyOn(productActions, "fetchProducts").mockImplementation(() => () => ({ type: "FETCH_PRODUCTS" }));

    wrapper = mount(
      <Provider store={store}>
        <AdminPanel />
      </Provider>
    );
  });

  afterEach(() => {
    jest.clearAllMocks();
  });

  it("renders AdminPanel without crashing", () => {
    expect(wrapper.find("h2").at(0).text()).toContain("Admin Panel - All Orders");
  });

  it("dispatches fetchAllOrders, fetchAllUsers, and fetchProducts on mount", () => {
    expect(orderActions.fetchAllOrders).toHaveBeenCalled();
    expect(userActions.fetchAllUsers).toHaveBeenCalled();
    expect(productActions.fetchProducts).toHaveBeenCalled();
  });

  it("shows loading message when loading is true", () => {
    const loadingState = {
      ...mockState,
      order: { ...mockState.order, loading: true },
    };
    store = createMockStore(loadingState);

    const loadingWrapper = mount(
      <Provider store={store}>
        <AdminPanel />
      </Provider>
    );

    expect(loadingWrapper.text()).toContain("Loading data...");
  });

  it("shows error message when an error is present", () => {
    const errorState = {
      ...mockState,
      order: { ...mockState.order, error: "Something went wrong" },
    };
    store = createMockStore(errorState);

    const errorWrapper = mount(
      <Provider store={store}>
        <AdminPanel />
      </Provider>
    );

    expect(errorWrapper.text()).toContain("Error: Something went wrong");
  });

  it("shows 'No Orders Found' when no orders exist", () => {
    const noOrdersState = {
      ...mockState,
      order: { ...mockState.order, allOrders: [] },
    };
    store = createMockStore(noOrdersState);

    const noOrdersWrapper = mount(
      <Provider store={store}>
        <AdminPanel />
      </Provider>
    );

    expect(noOrdersWrapper.text()).toContain("No Orders Found");
  });

  it("renders order and user tables correctly", () => {
    expect(wrapper.find("table").length).toBeGreaterThanOrEqual(2);

    // Verify order table content
    const orderRow = wrapper.find("tbody").at(0).find("tr").at(0);
    expect(orderRow.find("td").at(0).text()).toContain("#1234");
    expect(orderRow.find("td").at(4).text()).toContain("Cash on Delivery");
    expect(orderRow.find("td").at(5).text()).toContain("123 Main St");

    // Verify user table content
    const userRow = wrapper.find("tbody").at(1).find("tr").at(0);
    expect(userRow.find("td").at(0).text()).toContain("1");
    expect(userRow.find("td").at(1).text()).toContain("john_doe");
    expect(userRow.find("td").at(2).text()).toContain("john@example.com");
  });

  it("matches snapshot", () => {
    const snapshotWrapper = mount(
      <Provider store={store}>
        <AdminPanel />
      </Provider>
    );
    expect(snapshotWrapper).toMatchSnapshot();
  });
});