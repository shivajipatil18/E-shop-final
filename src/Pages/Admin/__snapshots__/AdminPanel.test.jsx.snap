// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AdminPanel Component matches snapshot 1`] = `ReactWrapper {}`;
